### PR TITLE
Azure Active Directory support

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
Adds support for authenticating and authorizing Azure AD users 

# Impacts
FreeRADIUS server (adds a new module in it)
PF authentication sources

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [X] Document the feature
- [X] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* Added support for Azure AD authentication

### CODE REVIEW

**Comment by julsemaan** [comment]
@extrafu, I included the perl module from https://github.com/jimdigriz/freeradius-oauth2-perl which is GPL licensed.

Not sure if this should be packaged as a separate dependency or if it can stay in our repo

The file in question is: raddb/mods-config/perl/oauth2.pm

File: docs/installation/azuread.asciidoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+You can perform 802.1x authentication of users using Azure AD but this will only work with supplicants configured to perform EAP-TTLS PAP which provides the RADIUS server with the plain-text password of the user. Support for this type of authentication is not as broad as EAP-PEAP MSCHAPv2 in the 802.1x supplicants but unfortunately Azure AD doesn't support MSCHAP authentication. Refer to the documentation of your operating system on how to configure EAP-TTLS PAP. This section will only focus on enabling EAP-TTLS PAP for your Azure AD users in PacketFence.
+
+ 1. Under 'Configuration->Policies and Access Control->Realms', create a new realm
+  1. Realm: enter the realm of your Azure AD users. Example, if the usernames have the following format `bob@inverseinc.onmicrosoft.com`, then your realm is `inverseinc.onmicrosoft.com`
+  1. Go in the 'Stripping' tab of the realm and select your Azure AD source under 'Azure AD Source for TTLS PAP'
+  1. Still in the 'Stripping' tab, disable (uncheck), the following settings:
+   1. Strip on the portal
+   1. Strip on the admin
+   1. Strip in RADIUS authorization
+  1. Save the realm
```
Comment (fdurand) [inline docs/installation/azuread.asciidoc:78]: Don't we have to set no strip in the freeradius proxy options ?

File: lib/pf/UnifiedApi/Controller/Config/Sources.pm [hunk 2]
```diff
@@ -27,8 +27,10 @@ use pfappserver::Form::Config::Source;
 use pfappserver::Form::Config::Source::AdminProxy;
 use pfappserver::Form::Config::Source::AD;
 use pfappserver::Form::Config::Source::AuthorizeNet;
+use pfappserver::Form::Config::Source::AzureAD;
 use pfappserver::Form::Config::Source::Blackhole;
 use pfappserver::Form::Config::Source::Authorization;
+use pfappserver::Form::Config::Source::AzureAD;
```
Comment (fdurand) [inline lib/pf/UnifiedApi/Controller/Config/Sources.pm]: duplicate

File: raddb/mods-config/perl/oauth2.pm [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+use LWP::UserAgent;
+use LWP::ConnCache;
+use POSIX qw(setlocale LC_ALL);
+use Time::Piece;
+
+use Data::Dumper;
+
+# https://wiki.freeradius.org/modules/Rlm_perl#logging is wrong...
+use constant {
+	L_AUTH			=> 2,
```
Comment (fdurand) [inline raddb/mods-config/perl/oauth2.pm]: You will have to fix the identation

**Review by fdurand — COMMENTED** [review]
Except few details all looks to be ok. (just need to test it after merge)
@satkunas what about the admin gui part ?

File: docs/installation/azuread.asciidoc (hunk 1 above)
Comment (julsemaan) [reply docs/installation/azuread.asciidoc:78]: Not according to my tests, I got the username unstripped in httpd.aaa with that configuration

**Comment by julsemaan** [comment]
Adjusted based on the comments and rebased the branch.

The admin looks good with my rebase with @satkunas' latest changes but I'll let him be the judge on whether or not something else needs to be adjusted in the frontend

**Comment by satkunas** [comment]
@julsemaan the ui portion looks good.

File: html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeAzureAD.vue [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  FormGroupTimeout,
+  FormGroupRealms,
+  FormGroupUserGroupsCache,
+}
+
+import { useForm as setup, useFormProps as props } from '../_composables/useForm'
+
+// @vue/component
+export default {
+  name: 'form-type-edir',
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeAzureAD.vue]: name: 'form-type-azure-a-d'

**Comment by julsemaan** [comment]
@extrafu, can you take a look at the doc

We'll also need to discuss where is the best place to put this in our install guide

File: docs/installation/azuread.asciidoc [hunk 5]
```diff
@@ -0,0 +1,83 @@
+
+TODO: This needs to be placed inside the installation guide
+
+== Azure AD integration
+
+PacketFence supports integrating with the Azure Active Directory for authenticating users on the captive-portal, the admin interface and for 802.1x users using EAP-TTLS PAP. If your only goal is to authenticate users on the captive-portal, using the OpenID implementation of Azure AD may be better suited. This section is aimed at providing username/password authentication through Azure AD.
```
Comment (extrafu) [inline docs/installation/azuread.asciidoc]: everywhere: 802.1x -> 802.1X

Comment (extrafu) [inline docs/installation/azuread.asciidoc]: everywhere:  captive-portal -> captive portal

File: raddb/mods-config/perl/oauth2.pm [hunk 6]
```diff
@@ -0,0 +1,400 @@
+use strict;
```
Comment (extrafu) [inline raddb/mods-config/perl/oauth2.pm:13]: We need to state (c) ownership on that file + license.

Comment (julsemaan) [reply raddb/mods-config/perl/oauth2.pm:13]: Just to make sure, would adding the following perldoc work:

```
=head1 COPYRIGHT

Copyright Alexander Clouter (https://github.com/jimdigriz)

Licence: https://github.com/jimdigriz/freeradius-oauth2-perl/blob/master/LICENSE

=cut
```

Comment (extrafu) [reply raddb/mods-config/perl/oauth2.pm:13]: Yes it should be OK.

**Comment by julsemaan** [comment]
PR adjusted, I think all that is left is to decide where to put the section in our doc and we're good for merging

**Comment by extrafu** [comment]
> PR adjusted, I think all that is left is to decide where to put the section in our doc and we're good for merging

I suggest we put it here: https://www.packetfence.org/doc/PacketFence_Installation_Guide.html#_authentication_mechanisms

after "External API Authentication"

**Comment by julsemaan** [comment]
Doc has been placed into the install guide, should be ready for merging AFAIK

**Comment by extrafu** [comment]
@fdurand Please merge and update the NEWS file!

File: raddb/mods-config/perl/oauth2.pm [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+
+Copyright Alexander Clouter (https://github.com/jimdigriz)
+
+Licence: https://github.com/jimdigriz/freeradius-oauth2-perl/blob/master/LICENSE
+
+=cut
+
+use strict;
+use warnings;
+
```
Comment (fdurand) [inline raddb/mods-config/perl/oauth2.pm:15]: use lib qw(
    /usr/local/pf/lib_perl/lib/perl5
);

Comment (julsemaan) [reply raddb/mods-config/perl/oauth2.pm:15]: Don

**Pull request merged**